sygnal/firebasepushkin.py

```diff
@@ -55,217 +76,233 @@ def __init__(self, name, sygnal, config):
+            "firebase_dispatch", tags=span_tags, child_of=context.opentracing_span
+        ) as span_parent:
+
+            dispatch_handler = self._map_event_dispatch_handler(n)
+            if dispatch_handler is None:
+                return []  # skipped
+
+            for retry_number in range(self.MAX_TRIES):
+                try:
+                    log.debug("Trying")
+
+                    span_tags = {"retry_num": retry_number}
+
+                    with self.sygnal.tracer.start_span(
+                        "firebase_dispatch_try", tags=span_tags, child_of=span_parent
+                    ) as span:
+                        return await dispatch_handler(device, span, log)
+                except TemporaryNotificationDispatchException as ex:
+                    retry_delay = self.RETRY_DELAY_BASE * (2 ** retry_number)
+                    if ex.custom_retry_delay is not None:
+                        retry_delay = ex.custom_retry_delay
+
+                    log.warning(
+                        "Temporary failure, will retry in %d seconds",
+                        retry_delay,
+                        exc_info=True,
+                    )
+                    span_parent.log_kv(
+                        {"event": "temporary_fail", "retrying_in": retry_delay}
+                    )
+                    if retry_number == self.MAX_TRIES - 1:
+                        raise NotificationDispatchException(
+                            "Retried too many times."
+                        ) from ex
+                    else:
+                        await twisted_sleep(
+                            retry_delay, twisted_reactor=self.sygnal.reactor
+                        )
 
-        return failed
+    @staticmethod
+    def _map_counts_unread(n):
+        return n.counts.unread or 0
 
-    async def send(self, message):
+    @staticmethod
+    def _map_android_priority(n):
+        return "normal" if n.prio == "low" else "high"
 
-        d = Deferred()
+    @staticmethod
+    def _map_ios_priority(n):
+        return "10" if n.prio == "high" else "5"
 
-        def done(success, result):
-            self.reactor.callFromThread(d.callback, result)
+    @staticmethod
+    def _message_body_from_notification(n, message_types):
+        if n.type != "m.room.message":
+            return ""
 
-        with SEND_TIME_HISTOGRAM.time():
-            self._pool.callInThreadWithCallback(
-                done, messaging.send_multicast, message, app=self._app
-            )
-            response = await d
+        from_display = ""
+        if n.room_name is not None and n.sender_display_name is not None:
+            from_display = n.sender_display_name + ": "
 
-        logger.debug(
-            "Message send success: %s of %s",
-            response.success_count,
-            response.success_count + response.failure_count,
-        )
+        body = message_types.get(n.content["msgtype"])
+        if body:
+            return body
```

Comment:
This should probably be moved above the `from_display` calculation, otherwise we can end up calculating `from_display` unnecessarily.